### ISSUE TITLE
aws-lambda-nodejs: Default entry finding fails when defining file starts with "file://" 

### ISSUE THREAD

**Issue by jonbarksdale** [aws/aws-cdk#21630]
### Describe the bug

Attempting to create a lambda with NodejsFunction, which should auto find the handler based on the defining files path.  

I get an error that looks like this: 

```
Error: Cannot find handler file file:///Users/stack.handler.ts, file:///Users/<path>/stack.handler.js or file:///Users/<path>/stack.handler.mjs
```

where the middle of the path has been replaced with <path> for brevity and privacy. 

[This](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-lambda-nodejs/lib/function.ts#L189) implies the paths being checked start with "file://", which I've tested and does not work with Node 16.13.1.

Those paths are produced from [this function](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-lambda-nodejs/lib/function.ts#L195-L211), which I believe is incorrect.  

This is running on macos 12.5, with the a fore mentioned version of nodejs. 

### Expected Behavior

I expected it to find the file, as the path was correct without the `file://` prefix.  

### Current Behavior

Error: Cannot find handler file file:///Users/stack.handler.ts, file:///Users/<path>/stack.handler.js or file:///Users/<path>/stack.handler.mjs

### Reproduction Steps

Create a lambda function using `NodejsFunction` with no `entry` parameter specified, and synthesize. 

### Possible Solution

`callsites()` produces paths that start with file:// when they are local files, and `findDefiningFile` doesn't scrub the path. Maybe make findDefining file remove the prefix when it's `file://`

### CDK CLI Version

2.37.1 (build f15dee0)

### Node.js Version

16.13.1

### OS

macos 12.5

### Language

Typescript

### Language Version

4.7.2

**Comment by corymhall** [comment]
@jonbarksdale I've tested this out on the same node version and macos version and have not been able to reproduce the issue. Can you try reproducing on a fresh project and let me know if you can reproduce?

**Comment by okko** [comment]
Workaround: https://github.com/sindresorhus/callsites/issues/18#issuecomment-1206281003

Node 16.17.0 was expected to fix this according to the callsites issue 18 above, https://github.com/nodejs/node/issues/39787 but I'm experiencing still the same with Node 16.17.0 when using ESM.

**Comment by okko** [comment]
@corymhall To reproduce please see https://github.com/okko/aws-cdk-esm-file-urls-reproduce

**Comment by corymhall** [comment]
@okko thanks for the additional info and the reproduction! I think I'm fine with the workaround that you suggest `replace('file://', '')`.

**Comment by jonbarksdale** [comment]
Thanks @okko, appreciate you stepping in with the reproduction.

**Comment by okko** [comment]
Workaround until issue is resolved in aws-cdk:
```
const handler = new lambdaNodejs.NodejsFunction(scope, functionId, {
  entry: new URL(import.meta.url.replace(/(.*)(\..+)/, '$1.' + functionId + '$2')).pathname, // https://github.com/aws/aws-cdk/pull/21802#issuecomment-1249940400
})
```

**Comment by e3dio** [comment]
So the default way to use Node.js Lambda function does not work, and no progress has been made for 11 months to fix it? I am getting same error

**Comment by nickej** [comment]
Is ` entry: new URL(import.meta.url.replace(/(.*)(\..+)/, '$1.' + functionId + '$2')).pathname` still the best workaround for this issue? It looks lite the PR by @okko was abandoned. I could look into making a new PR with the `replace('file://')` fix since this seems to still be an issue.

**Comment by okko** [comment]
Yes, it is still the best workaround. I appreciate if you @nickej  want to continue with the code from https://github.com/aws/aws-cdk/pull/21802/files. At the time it was blocked due to no tooling for ESM in the tests and @TheRealAmazonKendra requiring integration tests. The NodeJSFunction has been broken for ESM functions for over 2 years.

**Comment by nickej** [comment]
Ok, thanks for confirming @okko. I've worked a bit in my own [fork](https://github.com/nickej/aws-cdk/tree/fix_aws-lambda-nodejs_lib_function_esm_file_urls) on this but unfortunately it looks like there is still no easy way to run integration-tests using ESM. At least not something I found. Writing the test itself is easy enough but setting it up so that the code under test will be generated towards ESM is not something I had any luck with. 

I have asked for support in the cdk-dev Slack.

**Comment by Ys-Zhou** [comment]
It has been 2 years. Since node18 will be EOL in monthes, I really think cdk should support ESM by default and give a flag to anyone who wants to keep on CJS.

**Comment by okko** [comment]
@TheRealAmazonKendra Could you kindly reconsider this? The fix to aws-cdk would be simply `.replace('file://', '')` (https://github.com/aws/aws-cdk/pull/21802/files), doesn't break any existing functionality, and currently everyone using ESM still has to resort to the entry hack.
